Trouble with using ranges-v3 and hpx::parallel::reduce

This may be more appropriate for stack overflow, but... I was hoping to get some advice on how to convince hpx that it is safe to apply a parallel reduction to a `ranges::view::transform` that unfortunately winds up with an input iterator category (see detail below). 

I tried this:
```c++
#include <boost/iterator/iterator_categories.hpp>  
namespace boost{
template<>
struct iterator_traversal<IterR>{
  using type = random_access_traversal_tag;
};
}
```
where IterR is:

```c++ 
using R = decltype(ranges::view::repeat_n(1.0,1000000000)
		   | ranges::view::transform(add_one) 
		   | ranges::view::bounded);

using IterR = decltype(R{}.begin())
``` 

## Expected Behavior

I would like to pass the result of `ranges::view::transform` to `hpx::parallel::reduce` and successfully compile/execute.

## Actual Behavior

Fails to compile because per Eric Niebler:

"By design. The old standard iterator categories require forward iterators to return real references. Your transform_range returns prvalues; hence, its iterator can model (C++98) InputIterator at best.

However, if you ask for ranges::iterator_category_t, you'll get the new-style iterator category, which is random-access."

https://github.com/ericniebler/range-v3/issues/961

Compiler output:
```
hpx/include/hpx/include/parallel_reduce.hpp:10:
hpx/include/hpx/parallel/algorithms/reduce.hpp:107:13: error: static_assert failed due to requirement 'hpx::traits::is_forward_iterator<basic_iterator<adaptor_cursor<basic_iterator<cursor>, adaptor<true> >
      > >::value' "Requires at least forward iterator."
            static_assert(
            ^
hpx/include/hpx/parallel/algorithms/reduce.hpp:275:24: note: in instantiation of function template specialization 'hpx::parallel::v1::detail::reduce_<const hpx::parallel::execution::sequenced_policy &,
      ranges::v3::basic_iterator<ranges::v3::adaptor_cursor<ranges::v3::basic_iterator<ranges::v3::repeat_n_view<double>::cursor>, ranges::v3::iter_transform_view<ranges::v3::repeat_n_view<double>,
      ranges::v3::indirected<(lambda at hpx_example/hpx_example.cpp:3:22)> >::adaptor<true> > >, double, std::__1::plus<double> >' requested here
        return detail::reduce_(
                       ^
hpx_example/hpx_example.cpp:44:32: note: in instantiation of function template specialization 'hpx::parallel::v1::reduce<const hpx::parallel::execution::sequenced_policy &,
      ranges::v3::basic_iterator<ranges::v3::adaptor_cursor<ranges::v3::basic_iterator<ranges::v3::repeat_n_view<double>::cursor>, ranges::v3::iter_transform_view<ranges::v3::repeat_n_view<double>,
      ranges::v3::indirected<(lambda at hpx_example/hpx_example.cpp:3:22)> >::adaptor<true> > >, double>' requested here
  auto result = hpx::parallel::reduce(hpx::parallel::execution::seq,
```

## Steps to Reproduce the Problem

https://wandbox.org/permlink/KcWne4bEQVJWdwNJ

## Specifications

... Please describe your environment

  - HPX Version: 3.3.2
  - Platform (compiler, OS): mac llvm 6.0.1

--- a/hpx/parallel/algorithms/detail/accumulate.hpp
+++ b/hpx/parallel/algorithms/detail/accumulate.hpp
@@ -1,0 +1,33 @@
+//  Copyright (c) 2019 Hartmut Kaiser
+//
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#if !defined(HPX_PARALLEL_DETAIL_ACCUMULATE_2019_FEB_02_0527PM)
+#define HPX_PARALLEL_DETAIL_ACCUMULATE_2019_FEB_02_0527PM
+
+#include <hpx/config.hpp>
+
+#include <functional>
+
+namespace hpx { namespace parallel { inline namespace v1 { namespace detail
+{
+    // provide implementation of std::accumulate supporting iterators/sentinels
+    template <typename InIterB, typename InIterE, typename T, typename F>
+    inline T accumulate(InIterB first, InIterE last, T value, F reduce_op)
+    {
+        for (/**/; first != last; ++first)
+        {
+            value = reduce_op(value, *first);
+        }
+        return value;
+    }
+
+    template <typename InIterB, typename InIterE, typename T>
+    inline T accumulate(InIterB first, InIterE last, T value)
+    {
+        return accumulate(first, last, value, std::plus<>());
+    }
+}}}}
+
+#endif

--- a/hpx/parallel/algorithms/detail/accumulate.hpp
+++ b/hpx/parallel/algorithms/detail/accumulate.hpp
@@ -26,7 +26,7 @@ namespace hpx { namespace parallel { inline namespace v1 { namespace detail
     template <typename InIterB, typename InIterE, typename T>
     inline T accumulate(InIterB first, InIterE last, T value)
     {
-        return accumulate(first, last, value, std::plus<>());
+        return accumulate(first, last, value, std::plus<T>());
     }
 }}}}
 

--- a/hpx/parallel/algorithms/detail/distance.hpp
+++ b/hpx/parallel/algorithms/detail/distance.hpp
@@ -14,7 +14,7 @@ namespace hpx { namespace parallel { inline namespace v1 { namespace detail
 {
     // provide implementation of std::distance supporting iterators/sentinels
     template <typename InIterB, typename InIterE>
-    HPX_CONSTEXPR inline
+    HPX_CXX14_CONSTEXPR inline
     typename std::iterator_traits<InIterB>::difference_type
     distance(InIterB first, InIterE last, std::input_iterator_tag)
     {

--- a/hpx/parallel/algorithms/detail/distance.hpp
+++ b/hpx/parallel/algorithms/detail/distance.hpp
@@ -1,0 +1,47 @@
+//  Copyright (c) 2019 Hartmut Kaiser
+//
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#if !defined(HPX_PARALLEL_DETAIL_DISTANCE_2019_FEB_02_0508PM)
+#define HPX_PARALLEL_DETAIL_DISTANCE_2019_FEB_02_0508PM
+
+#include <hpx/config.hpp>
+
+#include <iterator>
+
+namespace hpx { namespace parallel { inline namespace v1 { namespace detail
+{
+    // provide implementation of std::distance supporting iterators/sentinels
+    template <typename InIterB, typename InIterE>
+    HPX_CONSTEXPR inline
+    typename std::iterator_traits<InIterB>::difference_type
+    distance(InIterB first, InIterE last, std::input_iterator_tag)
+    {
+        typename std::iterator_traits<InIterB>::difference_type offset = 0;
+        for (/**/; first != last; ++first)
+        {
+            ++offset;
+        }
+        return offset;
+    }
+
+    template <typename RanIterB, typename RanIterE>
+    HPX_CONSTEXPR inline
+    typename std::iterator_traits<RanIterB>::difference_type
+    distance(RanIterB first, RanIterE last, std::random_access_iterator_tag)
+    {
+        return last - first;
+    }
+
+    template <typename InIterB, typename InIterE>
+    HPX_CONSTEXPR inline
+    typename std::iterator_traits<InIterB>::difference_type
+    distance(InIterB first, InIterE last)
+    {
+        return distance(first, last,
+            typename std::iterator_traits<InIterB>::iterator_category{});
+    }
+}}}}
+
+#endif

--- a/hpx/parallel/algorithms/reduce.hpp
+++ b/hpx/parallel/algorithms/reduce.hpp
@@ -13,7 +13,9 @@
 #include <hpx/util/range.hpp>
 #include <hpx/util/unwrap.hpp>
 
+#include <hpx/parallel/algorithms/detail/accumulate.hpp>
 #include <hpx/parallel/algorithms/detail/dispatch.hpp>
+#include <hpx/parallel/algorithms/detail/distance.hpp>
 #include <hpx/parallel/execution_policy.hpp>
 #include <hpx/parallel/util/detail/algorithm_result.hpp>
 #include <hpx/parallel/util/loop.hpp>
@@ -41,21 +43,20 @@ namespace hpx { namespace parallel { inline namespace v1
               : reduce::algorithm("reduce")
             {}
 
-            template <typename ExPolicy, typename InIter, typename T_,
-                typename Reduce>
-            static T
-            sequential(ExPolicy, InIter first, InIter last, T_ && init,
-                Reduce && r)
+            template <typename ExPolicy, typename InIterB, typename InIterE,
+                typename T_, typename Reduce>
+            static T sequential(
+                ExPolicy, InIterB first, InIterE last, T_&& init, Reduce&& r)
             {
-                return std::accumulate(first, last, std::forward<T_>(init),
+                return detail::accumulate(first, last, std::forward<T_>(init),
                     std::forward<Reduce>(r));
             }
 
-            template <typename ExPolicy, typename FwdIter, typename T_,
-                typename Reduce>
+            template <typename ExPolicy, typename FwdIterB, typename FwdIterE,
+                typename T_, typename Reduce>
             static typename util::detail::algorithm_result<ExPolicy, T>::type
-            parallel(ExPolicy && policy, FwdIter first, FwdIter last,
-                T_ && init, Reduce && r)
+            parallel(ExPolicy&& policy, FwdIterB first, FwdIterE last,
+                T_&& init, Reduce&& r)
             {
                 if (first == last)
                 {
@@ -64,7 +65,7 @@ namespace hpx { namespace parallel { inline namespace v1
                 }
 
                 auto f1 =
-                    [r](FwdIter part_begin, std::size_t part_size) -> T
+                    [r](FwdIterB part_begin, std::size_t part_size) -> T
                     {
                         T val = *part_begin;
                         return util::accumulate_n(++part_begin, --part_size,
@@ -73,12 +74,11 @@ namespace hpx { namespace parallel { inline namespace v1
 
                 return util::partitioner<ExPolicy, T>::call(
                     std::forward<ExPolicy>(policy),
-                    first, std::distance(first, last),
+                    first, detail::distance(first, last),
                     std::move(f1),
                     hpx::util::unwrapping(
-                        [HPX_CAPTURE_FORWARD(init),
-                            HPX_CAPTURE_FORWARD(r)
-                        ](std::vector<T> && results) -> T
+                        [HPX_CAPTURE_FORWARD(init), HPX_CAPTURE_FORWARD(r)](
+                            std::vector<T> && results) -> T
                         {
                             return util::accumulate_n(hpx::util::begin(results),
                                 hpx::util::size(results), init, r);
@@ -87,25 +87,26 @@ namespace hpx { namespace parallel { inline namespace v1
         };
         /// \endcond
         // Non Segmented Reduce
-        template <typename ExPolicy, typename FwdIter, typename T, typename F>
+        template <typename ExPolicy, typename FwdIterB, typename FwdIterE,
+            typename T, typename F>
         inline typename std::enable_if<
             execution::is_execution_policy<ExPolicy>::value,
             typename util::detail::algorithm_result<ExPolicy, T>::type
         >::type
-        reduce_(ExPolicy&& policy, FwdIter first, FwdIter last
-            , T init, F && f, std::false_type)
+        reduce_(ExPolicy&& policy, FwdIterB first, FwdIterE last, T init, F&& f,
+            std::false_type)
         {
 #if defined(HPX_HAVE_ALGORITHM_INPUT_ITERATOR_SUPPORT)
             static_assert(
-                (hpx::traits::is_input_iterator<FwdIter>::value),
+                (hpx::traits::is_input_iterator<FwdIterB>::value),
                 "Requires at least input iterator.");
             typedef std::integral_constant<bool,
                     execution::is_sequenced_execution_policy<ExPolicy>::value ||
-                   !hpx::traits::is_forward_iterator<FwdIter>::value
+                   !hpx::traits::is_forward_iterator<FwdIterB>::value
                 > is_seq;
 #else
             static_assert(
-                (hpx::traits::is_forward_iterator<FwdIter>::value),
+                (hpx::traits::is_forward_iterator<FwdIterB>::value),
                 "Requires at least forward iterator.");
 
             typedef execution::is_sequenced_execution_policy<ExPolicy> is_seq;
@@ -117,13 +118,14 @@ namespace hpx { namespace parallel { inline namespace v1
         }
 
         // Forward Declaration of Segmented Reduce
-        template <typename ExPolicy, typename FwdIter, typename T, typename F>
+        template <typename ExPolicy, typename FwdIterB, typename FwdIterE,
+            typename T, typename F>
         inline typename std::enable_if<
             execution::is_execution_policy<ExPolicy>::value,
             typename util::detail::algorithm_result<ExPolicy, T>::type
         >::type
-        reduce_(ExPolicy&& policy, FwdIter first, FwdIter last
-            , T init, F && f, std::true_type);
+        reduce_(ExPolicy&& policy, FwdIterB first, FwdIterE last, T init, F&& f,
+            std::true_type);
     }
 
     /// Returns GENERALIZED_SUM(f, init, *first, ..., *(first + (last - first) - 1)).
@@ -135,7 +137,10 @@ namespace hpx { namespace parallel { inline namespace v1
     ///                     It describes the manner in which the execution
     ///                     of the algorithm may be parallelized and the manner
     ///                     in which it executes the assignments.
-    /// \tparam FwdIter     The type of the source iterators used (deduced).
+    /// \tparam FwdIterB    The type of the source begin iterator used (deduced).
+    ///                     This iterator type must meet the requirements of an
+    ///                     forward iterator.
+    /// \tparam FwdIterE    The type of the source end iterator used (deduced).
     ///                     This iterator type must meet the requirements of an
     ///                     forward iterator.
     /// \tparam F           The type of the function/function object to use
@@ -161,7 +166,7 @@ namespace hpx { namespace parallel { inline namespace v1
     ///                     \endcode \n
     ///                     The signature does not need to have const&.
     ///                     The types \a Type1 \a Ret must be
-    ///                     such that an object of type \a FwdIter can be
+    ///                     such that an object of type \a FwdIterB can be
     ///                     dereferenced and then implicitly converted to any
     ///                     of those types.
     /// \param init         The initial value for the generalized sum.
@@ -196,14 +201,15 @@ namespace hpx { namespace parallel { inline namespace v1
     /// that the behavior of reduce may be non-deterministic for
     /// non-associative or non-commutative binary predicate.
     ///
-    template <typename ExPolicy, typename FwdIter, typename T, typename F>
+    template <typename ExPolicy, typename FwdIterB, typename FwdIterE,
+        typename T, typename F>
     inline typename std::enable_if<
         execution::is_execution_policy<ExPolicy>::value,
         typename util::detail::algorithm_result<ExPolicy, T>::type
     >::type
-    reduce(ExPolicy&& policy, FwdIter first, FwdIter last, T init, F && f)
+    reduce(ExPolicy&& policy, FwdIterB first, FwdIterE last, T init, F&& f)
     {
-        typedef hpx::traits::is_segmented_iterator<FwdIter> is_segmented;
+        typedef hpx::traits::is_segmented_iterator<FwdIterB> is_segmented;
 
         return detail::reduce_(
             std::forward<ExPolicy>(policy), first, last,
@@ -219,7 +225,10 @@ namespace hpx { namespace parallel { inline namespace v1
     ///                     It describes the manner in which the execution
     ///                     of the algorithm may be parallelized and the manner
     ///                     in which it executes the assignments.
-    /// \tparam FwdIter     The type of the source iterators used (deduced).
+    /// \tparam FwdIterB    The type of the source begin iterator used (deduced).
+    ///                     This iterator type must meet the requirements of an
+    ///                     forward iterator.
+    /// \tparam FwdIterE    The type of the source end iterator used (deduced).
     ///                     This iterator type must meet the requirements of an
     ///                     forward iterator.
     /// \tparam T           The type of the value to be used as initial (and
@@ -263,14 +272,15 @@ namespace hpx { namespace parallel { inline namespace v1
     /// that the behavior of reduce may be non-deterministic for
     /// non-associative or non-commutative binary predicate.
     ///
-    template <typename ExPolicy, typename FwdIter, typename T>
+    template <typename ExPolicy, typename FwdIterB, typename FwdIterE,
+        typename T>
     inline typename std::enable_if<
         execution::is_execution_policy<ExPolicy>::value,
         typename util::detail::algorithm_result<ExPolicy, T>::type
     >::type
-    reduce(ExPolicy&& policy, FwdIter first, FwdIter last, T init)
+    reduce(ExPolicy&& policy, FwdIterB first, FwdIterE last, T init)
     {
-        typedef hpx::traits::is_segmented_iterator<FwdIter> is_segmented;
+        typedef hpx::traits::is_segmented_iterator<FwdIterB> is_segmented;
 
         return detail::reduce_(
             std::forward<ExPolicy>(policy), first, last,
@@ -286,7 +296,10 @@ namespace hpx { namespace parallel { inline namespace v1
     ///                     It describes the manner in which the execution
     ///                     of the algorithm may be parallelized and the manner
     ///                     in which it executes the assignments.
-    /// \tparam FwdIter     The type of the source iterators used (deduced).
+    /// \tparam FwdIterB    The type of the source begin iterator used (deduced).
+    ///                     This iterator type must meet the requirements of an
+    ///                     forward iterator.
+    /// \tparam FwdIterE    The type of the source end iterator used (deduced).
     ///                     This iterator type must meet the requirements of an
     ///                     forward iterator.
     ///
@@ -312,13 +325,13 @@ namespace hpx { namespace parallel { inline namespace v1
     ///           \a sequenced_task_policy or
     ///           \a parallel_task_policy and
     ///           returns T otherwise (where T is the value_type of
-    ///           \a FwdIter).
+    ///           \a FwdIterB).
     ///           The \a reduce algorithm returns the result of the
     ///           generalized sum (applying operator+()) over the elements given
     ///           by the input range [first, last).
     ///
     /// \note   The type of the initial value (and the result type) \a T is
-    ///         determined from the value_type of the used \a FwdIter.
+    ///         determined from the value_type of the used \a FwdIterB.
     ///
     /// \note   GENERALIZED_SUM(+, a1, ..., aN) is defined as follows:
     ///         * a1 when N is 1
@@ -331,18 +344,18 @@ namespace hpx { namespace parallel { inline namespace v1
     /// that the behavior of reduce may be non-deterministic for
     /// non-associative or non-commutative binary predicate.
     ///
-    template <typename ExPolicy, typename FwdIter>
+    template <typename ExPolicy, typename FwdIterB, typename FwdIterE>
     inline typename std::enable_if<
         execution::is_execution_policy<ExPolicy>::value,
         typename util::detail::algorithm_result<ExPolicy,
-            typename std::iterator_traits<FwdIter>::value_type
+            typename std::iterator_traits<FwdIterB>::value_type
         >::type
     >::type
-    reduce(ExPolicy&& policy, FwdIter first, FwdIter last)
+    reduce(ExPolicy&& policy, FwdIterB first, FwdIterE last)
     {
-        typedef typename std::iterator_traits<FwdIter>::value_type value_type;
+        typedef typename std::iterator_traits<FwdIterB>::value_type value_type;
 
-        typedef hpx::traits::is_segmented_iterator<FwdIter> is_segmented;
+        typedef hpx::traits::is_segmented_iterator<FwdIterB> is_segmented;
 
         return detail::reduce_(
             std::forward<ExPolicy>(policy), first, last,

--- a/hpx/parallel/segmented_algorithms/reduce.hpp
+++ b/hpx/parallel/segmented_algorithms/reduce.hpp
@@ -10,7 +10,9 @@
 #include <hpx/traits/segmented_iterator_traits.hpp>
 #include <hpx/util/invoke.hpp>
 
+#include <hpx/parallel/algorithms/detail/accumulate.hpp>
 #include <hpx/parallel/algorithms/detail/dispatch.hpp>
+#include <hpx/parallel/algorithms/detail/distance.hpp>
 #include <hpx/parallel/algorithms/reduce.hpp>
 #include <hpx/parallel/execution_policy.hpp>
 #include <hpx/parallel/segmented_algorithms/detail/dispatch.hpp>
@@ -37,14 +39,14 @@ namespace hpx { namespace parallel { inline namespace v1
         /// \cond NOINTERNAL
 
         // sequential remote implementation
-        template <typename Algo, typename ExPolicy, typename SegIter,
-            typename T, typename Reduce>
+        template <typename Algo, typename ExPolicy, typename SegIterB,
+            typename SegIterE, typename T, typename Reduce>
         static typename util::detail::algorithm_result<ExPolicy, T>::type
         segmented_reduce(Algo && algo, ExPolicy const& policy,
-            SegIter first, SegIter last, T && init,
+            SegIterB first, SegIterE last, T && init,
             Reduce && red_op, std::true_type)
         {
-            typedef hpx::traits::segmented_iterator_traits<SegIter> traits;
+            typedef hpx::traits::segmented_iterator_traits<SegIterB> traits;
             typedef typename traits::segment_iterator segment_iterator;
             typedef typename traits::local_iterator local_iterator_type;
             typedef util::detail::algorithm_result<ExPolicy, T> result;
@@ -108,27 +110,27 @@ namespace hpx { namespace parallel { inline namespace v1
         }
 
         // parallel remote implementation
-        template <typename Algo, typename ExPolicy, typename SegIter,
-            typename T, typename Reduce>
+        template <typename Algo, typename ExPolicy, typename SegIterB,
+            typename SegIterE, typename T, typename Reduce>
         static typename util::detail::algorithm_result<ExPolicy, T>::type
         segmented_reduce(Algo && algo, ExPolicy const& policy,
-            SegIter first, SegIter last, T && init,
+            SegIterB first, SegIterE last, T && init,
             Reduce && red_op, std::false_type)
         {
-            typedef hpx::traits::segmented_iterator_traits<SegIter> traits;
+            typedef hpx::traits::segmented_iterator_traits<SegIterB> traits;
             typedef typename traits::segment_iterator segment_iterator;
             typedef typename traits::local_iterator local_iterator_type;
             typedef util::detail::algorithm_result<ExPolicy, T> result;
 
             typedef std::integral_constant<bool,
-                    !hpx::traits::is_forward_iterator<SegIter>::value
+                    !hpx::traits::is_forward_iterator<SegIterB>::value
                 > forced_seq;
 
             segment_iterator sit = traits::segment(first);
             segment_iterator send = traits::segment(last);
 
             std::vector<shared_future<T> > segments;
-            segments.reserve(std::distance(sit, send));
+            segments.reserve(detail::distance(sit, send));
 
             if (sit == send)
             {
@@ -196,7 +198,7 @@ namespace hpx { namespace parallel { inline namespace v1
                         >::call(r, errors);
 
                         // VS2015RC bails out if red_op is capture by ref
-                        return std::accumulate(
+                        return detail::accumulate(
                             r.begin(), r.end(), init,
                             [=](T const& val, shared_future<T>& curr)
                             {
@@ -208,12 +210,13 @@ namespace hpx { namespace parallel { inline namespace v1
 
         ///////////////////////////////////////////////////////////////////////
         // segmented implementation
-        template <typename ExPolicy, typename InIter, typename T, typename F>
+        template <typename ExPolicy, typename InIterB, typename InIterE,
+            typename T, typename F>
         inline typename std::enable_if<
             execution::is_execution_policy<ExPolicy>::value,
             typename util::detail::algorithm_result<ExPolicy, T>::type
         >::type
-        reduce_(ExPolicy&& policy, InIter first, InIter last,
+        reduce_(ExPolicy&& policy, InIterB first, InIterE last,
             T init, F && f, std::true_type)
         {
             typedef parallel::execution::is_sequenced_execution_policy<
@@ -235,12 +238,13 @@ namespace hpx { namespace parallel { inline namespace v1
         }
 
         // forward declare the non-segmented version of this algorithm
-        template <typename ExPolicy, typename InIter, typename T, typename F>
+        template <typename ExPolicy, typename InIterB, typename InIterE,
+            typename T, typename F>
         inline typename std::enable_if<
             execution::is_execution_policy<ExPolicy>::value,
             typename util::detail::algorithm_result<ExPolicy, T>::type
         >::type
-        reduce_(ExPolicy&& policy, InIter first, InIter last
+        reduce_(ExPolicy&& policy, InIterB first, InIterE last
             , T init, F && f, std::false_type);
 
         /// \endcond

--- a/tests/regressions/parallel/CMakeLists.txt
+++ b/tests/regressions/parallel/CMakeLists.txt
@@ -7,6 +7,7 @@ set(tests
     for_each_annotated_function
     for_loop_2281
     minimal_findend
+    reduce_3641
     scan_different_inits
     scan_non_commutative
     scan_shortlength

--- a/tests/regressions/parallel/reduce_3641.cpp
+++ b/tests/regressions/parallel/reduce_3641.cpp
@@ -8,8 +8,8 @@
 // #3646: Parallel algorithms should accept iterator/sentinel pairs
 
 #include <hpx/hpx_main.hpp>
-#include <hpx/util/lightweight_test.hpp>
 #include <hpx/include/parallel_reduce.hpp>
+#include <hpx/util/lightweight_test.hpp>
 
 #include <cstddef>
 #include <cstdint>
@@ -23,8 +23,7 @@ struct Iterator
 {
     using difference_type = std::ptrdiff_t;
     using value_type = std::int64_t;
-    // for now - just to verify reduce algorithm works with sentinels
-    using iterator_category = std::forward_iterator_tag;
+    using iterator_category = std::input_iterator_tag;
     using pointer = std::int64_t const*;
     using reference = std::int64_t const&;
 
@@ -36,56 +35,58 @@ struct Iterator
     }
     std::int64_t operator->() const = delete;
 
-    auto& operator++()
+    Iterator& operator++()
     {
         ++(this->state);
         return *this;
     }
 
-    auto operator++(int) const
+    Iterator operator++(int)
     {
         auto copy = *this;
-        return ++copy;
+        ++(*this);
+        return copy;
     }
 
-    auto& operator--()
+    Iterator& operator--()
     {
         --(this->state);
         return *this;
     }
 
-    auto operator--(int)
+    Iterator operator--(int)
     {
         auto copy = *this;
-        return --copy;
+        --(*this);
+        return copy;
     }
 
-    auto operator[](difference_type n) const
+    std::int64_t operator[](difference_type n) const
     {
         return this->state + n;
     }
 
-    auto& operator+=(difference_type n)
+    Iterator& operator+=(difference_type n)
     {
         this->state += n;
         return *this;
     }
 
-    auto operator+(difference_type n) const
+    Iterator operator+(difference_type n) const
     {
-        auto copy = *this;
+        Iterator copy = *this;
         return copy += n;
     }
 
-    auto& operator-=(difference_type n)
+    Iterator& operator-=(difference_type n)
     {
         this->state -= n;
         return *this;
     }
 
-    auto operator-(difference_type n) const
+    Iterator operator-(difference_type n) const
     {
-        auto copy = *this;
+        Iterator copy = *this;
         return copy -= n;
     }
 
@@ -140,7 +141,7 @@ struct Iterator
 
 int main()
 {
-    auto result = hpx::parallel::reduce(hpx::parallel::execution::seq,
+    std::int64_t result = hpx::parallel::reduce(hpx::parallel::execution::seq,
         Iterator{0}, Sentinel{}, std::int64_t(0));
 
     HPX_TEST_EQ(result, std::int64_t(4950));

--- a/tests/regressions/parallel/reduce_3641.cpp
+++ b/tests/regressions/parallel/reduce_3641.cpp
@@ -1,0 +1,154 @@
+//  Copyright (c) 2019 Austin McCartney
+//  Copyright (c) 2019 Hartmut Kaiser
+//
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+// #3641: Trouble with using ranges-v3 and hpx::parallel::reduce
+// #3646: Parallel algorithms should accept iterator/sentinel pairs
+
+#include <hpx/hpx_main.hpp>
+#include <hpx/util/lightweight_test.hpp>
+#include <hpx/include/parallel_reduce.hpp>
+
+#include <cstddef>
+#include <cstdint>
+#include <iterator>
+
+struct Sentinel
+{
+};
+
+struct Iterator
+{
+    using difference_type = std::ptrdiff_t;
+    using value_type = std::int64_t;
+    // for now - just to verify reduce algorithm works with sentinels
+    using iterator_category = std::forward_iterator_tag;
+    using pointer = std::int64_t const*;
+    using reference = std::int64_t const&;
+
+    std::int64_t state;
+
+    std::int64_t operator*() const
+    {
+        return this->state;
+    }
+    std::int64_t operator->() const = delete;
+
+    auto& operator++()
+    {
+        ++(this->state);
+        return *this;
+    }
+
+    auto operator++(int) const
+    {
+        auto copy = *this;
+        return ++copy;
+    }
+
+    auto& operator--()
+    {
+        --(this->state);
+        return *this;
+    }
+
+    auto operator--(int)
+    {
+        auto copy = *this;
+        return --copy;
+    }
+
+    auto operator[](difference_type n) const
+    {
+        return this->state + n;
+    }
+
+    auto& operator+=(difference_type n)
+    {
+        this->state += n;
+        return *this;
+    }
+
+    auto operator+(difference_type n) const
+    {
+        auto copy = *this;
+        return copy += n;
+    }
+
+    auto& operator-=(difference_type n)
+    {
+        this->state -= n;
+        return *this;
+    }
+
+    auto operator-(difference_type n) const
+    {
+        auto copy = *this;
+        return copy -= n;
+    }
+
+    bool operator==(const Iterator& that) const
+    {
+        return this->state == that.state;
+    }
+
+    friend bool operator==(Iterator i, Sentinel)
+    {
+        return i.state == 100;
+    }
+    friend bool operator==(Sentinel, Iterator i)
+    {
+        return i.state == 100;
+    }
+
+    bool operator!=(const Iterator& that) const
+    {
+        return this->state != that.state;
+    }
+
+    friend bool operator!=(Iterator i, Sentinel)
+    {
+        return i.state != 100;
+    }
+    friend bool operator!=(Sentinel, Iterator i)
+    {
+        return i.state != 100;
+    }
+
+    bool operator<(const Iterator& that) const
+    {
+        return this->state < that.state;
+    }
+
+    bool operator<=(const Iterator& that) const
+    {
+        return this->state <= that.state;
+    }
+
+    bool operator>(const Iterator& that) const
+    {
+        return this->state > that.state;
+    }
+
+    bool operator>=(const Iterator& that) const
+    {
+        return this->state >= that.state;
+    }
+};
+
+int main()
+{
+    auto result = hpx::parallel::reduce(hpx::parallel::execution::seq,
+        Iterator{0}, Sentinel{}, std::int64_t(0));
+
+    HPX_TEST_EQ(result, std::int64_t(4950));
+
+    result = hpx::parallel::reduce(hpx::parallel::execution::par,
+        Iterator{0}, Sentinel{}, std::int64_t(0));
+
+    HPX_TEST_EQ(result, std::int64_t(4950));
+
+    return hpx::util::report_errors();
+}

